devito/ir/clusters/cluster.py

```diff
@@ -344,11 +344,15 @@ def dspace(self):
                 else:
                     d = i.dim
                 try:
-                    if i.lower < 0 or \
-                       i.upper > f._size_nodomain[d].left + f._size_halo[d].right:
-                        # It'd mean trying to access a point before the
-                        # left halo (test0) or after the right halo (test1)
-                        oobs.update(d._defines)
+                    # TODO: This is the one check that needs to be turned off
+                    # for PETScArrays otherwise the iteration bounds are incorrect.
+                    # Need to discuss this.
+                    if not isinstance(f, PETScArray):
```

Comment:
couple of things here...

if the check is on `f`, it could be right at the top of `for f, v in parts` loop. 

I also think we can generalize it to ArrayBasic